File: java-sdks/src/main/java/com/suisdks/sui/deepbook_v3/DeepBookQueryOperations.java
package com.suisdks.sui.deepbook_v3;

import java.math.BigInteger;
import java.util.*;

public class DeepBookQueryOperations {
    private final DeepBookV3Client client;

    public DeepBookQueryOperations(DeepBookV3Client client) {
        this.client = client;
    }

    // ==================== Order Book Queries ====================

    public Map<String, Object> getLevel2Range(String sender, String poolKey,
            BigInteger priceLow, BigInteger priceHigh, boolean isBid) {
        Map<String, Object> txData = new HashMap<>();
        txData.put("kind", "programmableTransaction");
        
        Map<String, Object> moveCall = new HashMap<>();
        moveCall.put("kind", "moveCall");
        moveCall.put("target", client.config().deepbookPackage() + "::deepbook::get_level2_range");
        
        Map<String, Object> args = new HashMap<>();
        args.put("poolKey", poolKey);
        args.put("priceLow", priceLow.toString());
        args.put("priceHigh", priceHigh.toString());
        args.put("isBid", isBid);
        moveCall.put("arguments", List.of(args));
        
        txData.put("inputs", List.of());
        txData.put("transactions", List.of(moveCall));

        return client.callRpc("suix_devInspectTransactionBlock", List.of(sender, txData));
    }

    public Map<String, Object> getLevel2TicksFromMid(String sender, String poolKey,
            int ticks, boolean isBid) {
        Map<String, Object> txData = new HashMap<>();
        txData.put("kind", "programmableTransaction");
        
        Map<String, Object> moveCall = new HashMap<>();
        moveCall.put("kind", "moveCall");
        moveCall.put("target", client.config().deepbookPackage() + "::deepbook::get_level2_ticks_from_mid");
        
        Map<String, Object> args = new HashMap<>();
        args.put("poolKey", poolKey);
        args.put("ticks", ticks);
        args.put("isBid", isBid);
        moveCall.put("arguments", List.of(args));
        
        txData.put("inputs", List.of());
        txData.put("transactions", List.of(moveCall));

        return client.callRpc("suix_devInspectTransactionBlock", List.of(sender, txData));
    }

    public Map<String, Object> accountOpenOrders(String sender, String poolKey, String managerKey) {
        Map<String, Object> txData = new HashMap<>();
        txData.put("kind", "programmableTransaction");
        
        Map<String, Object> moveCall = new HashMap<>();
        moveCall.put("kind", "moveCall");
        moveCall.put("target", client.config().deepbookPackage() + "::deepbook::account_open_orders");
        
        Map<String, Object> args = new HashMap<>();
        args.put("poolKey", poolKey);
        args.put("managerKey", managerKey);
        moveCall.put("arguments", List.of(args));
        
        txData.put("inputs", List.of());
        txData.put("transactions", List.of(moveCall));

        return client.callRpc("suix_devInspectTransactionBlock", List.of(sender, txData));
    }

    public Map<String, Object> getOrder(String accountKey) {
        Map<String, Object> options = new HashMap<>();
        options.put("showContent", true);
        
        return client.callRpc("sui_getObject", Map.of(
            "objectId", accountKey,
            "options", options
        ));
    }

    public Map<String, Object> getOrders(List<String> accountKeys) {
        Map<String, Object> options = new HashMap<>();
        options.put("showContent", true);
        
        return client.callRpc("sui_multiGetObjects", List.of(accountKeys, options));
    }
}

File: swift-sdks/Sources/SuiSwiftSDK/DeepBookV3/DeepBookQueryOperations.swift
import Foundation

// Order Book Query Operations for Swift SDK
public class DeepBookQueryOperations {
    weak var client: DeepBookV3Client?

    public init(client: DeepBookV3Client) {
        self.client = client
    }

    // MARK: - Order Book Queries

    public func getLevel2Range(
        sender: String,
        poolKey: String,
        priceLow: UInt64,
        priceHigh: UInt64,
        isBid: Bool
    ) async throws -> [String: Any] {
        let txData = buildTransaction(
            target: "\((client?.config.deepbookPackage ?? "") + "::deepbook::get_level2_range")",            arguments: [
                ["poolKey": poolKey, "priceLow": priceLow, "priceHigh": priceHigh, "isBid": isBid]
            ]
        )

        return try await call("suix_devInspectTransactionBlock", params: [sender, txData])
    }

    public func getLevel2TicksFromMid(
        sender: String,
        poolKey: String,
        ticks: Int,
        isBid: Bool
    ) async throws -> [String: Any] {
        let txData = buildTransaction(
            target: "\((client?.config.deepbookPackage ?? "") + "::deepbook::get_level2_ticks_from_mid")",
            arguments: [
                ["poolKey": poolKey, "ticks": ticks, "isBid": isBid]
            ]
        )

        return try await call("suix_devInspectTransactionBlock", params: [sender, txData])
    }

    public func accountOpenOrders(
        sender: String,
        poolKey: String,
        managerKey: String
    ) async throws -> [String: Any] {
        let txData = buildTransaction(
            target: "\((client?.config.deepbookPackage ?? "") + "::deepbook::account_open_orders")",
            arguments: [
                ["poolKey": poolKey, "managerKey": managerKey]
            ]
        )

        return try await call("suix_devInspectTransactionBlock", params: [sender, txData])
    }

    public func getOrder(accountKey: String) async throws -> [String: Any] {
        let options: [String: Any] = ["showContent": true]
        
        return try await call("sui_getObject", params: [
            "objectId": accountKey,
            "options": options
        ])
    }

    public func getOrders(accountKeys: [String]) async throws -> [String: Any] {
        let options: [String: Any] = ["showContent": true]
        
        return try await call("sui_multiGetObjects", params: [accountKeys, options])
    }

    // MARK: - Helper Methods

    private func call(_ method: String, params params: [Any?] = []) async throws -> [String: Any] {
        guard let client = client else {
            throw DeepBookError.invalidResponse
        }
        return try await client.call(method, params: params)
    }

    private func buildTransaction(target: String, arguments: [[String: Any]]) -> [String: Any] {
        var transactions: [[String: Any]] = []
        
        for args in arguments {
            var moveCall: [String: Any] = [:]
            moveCall["kind"] = "moveCall"
            moveCall["target"] = target
            moveCall["arguments"] = args
            transactions.append(moveCall)
        }

        return [
            "kind": "programmableTransaction",
            "inputs": [],
            "transactions": transactions
        ]
    }
}